### PR TITLE
fix: unable to swipe b/w activity center tabs

### LINKED ISSUE #19000: Unable to swipe between tabs in the Activity Center; Tabs disappear after notifications are read/unread.
### Steps:
1. Open Activity Center (AC)
2. Select the last tab
3. Reopen AC
4. Attempt to swipe previous tabs 
5. Mark as read the notifications

### Actual result:
- In step 4: Only the last tab is shown in the AC, and the user is unable to select other previous tabs from the AC.

https://github.com/status-im/status-mobile/assets/52490791/917722a2-9e04-415c-9566-cf1177fc96ce

- In step 5: The tabs disappear in the AC after some messages are marked as read/unread.

https://github.com/status-im/status-mobile/assets/52490791/026ab2b2-daa2-483f-b0ee-1094a6fba321

### Expected result:
- The user should be able to swipe between tabs and select other tabs in the AC.
- The tabs within AC are not disappear after the messages are marked as read/unread

### Additional info:
The same issue occurs when other tabs are selected and the app is reopened. However, after swiping right and left, the previous tabs can be selected

### ENV:
Nightly [26 Feb 2024]

### PR DESCRIPTION
**Pull request by ajayesivan**
fixes #19000 

### Testing notes
- The fix is implemented in the tabs component, so ideally, we should test all scrollable tab usages.
- I was not able to reproduce the issue where the tabs disappear after marking notifications as read/unread. In case that issue is still reproducible in this pr build, please let me know.

status: ready

### CODE REVIEW

File: src/quo/components/tabs/tabs/view.cljs [hunk 1]
```diff
@@ -158,15 +158,12 @@
              :on-change
              :scroll-on-press?
              :size)
-            (when scroll-on-press?
-              {:initial-scroll-index (utils.collection/first-index #(= @active-tab-id (:id %)) data)})
```
Comment (ajayesivan) [inline src/quo/components/tabs/tabs/view.cljs]: While using `initial-scroll-index` without adding a `get-item-layout` function, it results in buggy behavior. In this case, implementing `get-item-layout` is not possible because the item width can't be identified beforehand, as the width depends on the tab title.

File: src/quo/components/tabs/tabs/view.cljs [hunk 2]
```diff
@@ -180,6 +177,15 @@
...
                                   :on-scroll           on-scroll})
+             :on-layout (fn []
+                          (js/setTimeout (fn []
+                                           (.scrollToIndex @flat-list-ref
+                                                           #js
+                                                            {:animated false
+                                                             :index    (utils.collection/first-index
+                                                                        #(= @active-tab-id (:id %))
+                                                                        data)}))
+                                         1))
```
Comment (ajayesivan) [inline src/quo/components/tabs/tabs/view.cljs]: Alternatively, I'm using the `scrollToIndex` to set the initial scroll position. Without `setTimeout`, this approach fails about 20% of the time. As far as I've tested with the timeout, it works reliably.

File: src/quo/components/tabs/tabs/view.cljs [hunk 3]
```diff
@@ -180,6 +177,15 @@
                                   :fade-end?           fade-end?
                                   :fading              fading
                                   :on-scroll           on-scroll})
+             :on-layout (fn []
```
Comment (flexsurfer) [inline src/quo/components/tabs/tabs/view.cljs]: please use `rn/use-callback` hook

Comment (flexsurfer) [reply src/quo/components/tabs/tabs/view.cljs]: for that, you need to refactor the entire component, and remove reagent render, feel free to ping me in Discord if you need help

**Comment by ajayesivan** [comment]
Hi @flexsurfer, I have updated the component to follow the [ui-guidelines](https://github.com/status-im/status-mobile/blob/develop/doc/ui-guidelines.md). Could you please have another look?

File: src/quo/components/tabs/tabs/view.cljs [hunk 4]
```diff
@@ -118,87 +119,99 @@
...
+                                :fading              fading
+                                :set-fading          set-fading
+                                :on-scroll           on-scroll})
+           :on-layout set-initial-scroll-poisition
+           :render-fn tab-view
+           :render-data {:active-tab-id       active-tab-id
+                         :set-active-tab-id   set-active-tab-id
+                         :blur?               blur?
+                         :customization-color customization-color
+                         :flat-list-ref       @flat-list-ref
```
Comment (flexsurfer) [inline src/quo/components/tabs/tabs/view.cljs]: `flat-list-ref` should be a state then, because ref won't re-render

File: src/quo/components/tabs/tabs/view.cljs [hunk 5]
```diff
@@ -72,31 +71,33 @@
...
+                                    (set-active-tab-id tab-id)
+                                    (when (and scroll-on-press? flat-list-ref)
+                                      (.scrollToIndex ^js flat-list-ref
+                                                      #js
+                                                       {:animated     true
+                                                        :index        index
+                                                        :viewPosition 0.5}))
+                                    (when on-change
+                                      (on-change tab-id)))
+                                  [flat-list-ref])]
```
Comment (flexsurfer) [inline src/quo/components/tabs/tabs/view.cljs]: or actually, seems it's better to move this function outside, then you can keep ref for flat-list-ref

File: src/quo/components/tabs/tabs/view.cljs [hunk 6]
```diff
@@ -118,87 +119,99 @@
...
+           in-scroll-view?
+           customization-color]
+    :or   {fade-end-percentage 0.8
+           fade-end?           false
+           scrollable?         false
+           scroll-on-press?    false
+           size                default-tab-size}
+    :as   props}]
+  (let [[active-tab-id set-active-tab-id] (rn/use-state default-active)
+        [fading set-fading]               (rn/use-state {:fade-end-percentage fade-end-percentage})
```
Comment (clauxx) [inline src/quo/components/tabs/tabs/view.cljs]: why is this a map?

File: src/quo/components/tabs/tabs/view.cljs [hunk 7]
```diff
@@ -118,87 +119,99 @@
...
+           :shows-horizontal-scroll-indicator false
+           :data data
+           :key-fn (comp str :id)
+           :on-scroll-to-index-failed identity
+           :on-scroll (partial on-scroll-handler
+                               {:fade-end-percentage fade-end-percentage
+                                :fade-end?           fade-end?
+                                :fading              fading
+                                :set-fading          set-fading
+                                :on-scroll           on-scroll})
```
Comment (clauxx) [inline src/quo/components/tabs/tabs/view.cljs]: This looks a bit confusing. We're passing ` fade-end-percentage` and `fading` state, which is just `{:fade-end-percentage  fade-end-percentage}`. Maybe `on-scroll-handler` could be a custom-hook that contains and updates the `fading` state, which would be returned together with the scroll handler e.g.

```clj
(let [...
    ;; `fading` state would be inside the hook instead of passing both the state and the setter to
    ;; `on-scroll-handler`, as it seems like no other place updates it
    {:keys [updated-fade-end-percentage on-scroll-handler]
         (use-tab-scroll {:fade-end-percentage fade-end-percentage
                          ...})}]
...)
```

File: src/quo/components/tabs/tabs/view.cljs [hunk 8]
```diff
@@ -118,87 +119,99 @@
...
+    :as   props}]
+  (let [[active-tab-id set-active-tab-id] (rn/use-state default-active)
+        [fading set-fading]               (rn/use-state {:fade-end-percentage fade-end-percentage})
+        flat-list-ref                     (rn/use-ref-atom nil)
+        set-flat-list-ref                 (rn/use-callback #(reset! flat-list-ref %))
+        set-initial-scroll-poisition      (rn/use-callback (fn []
+                                                             (js/setTimeout
+                                                              (fn []
+                                                                (.scrollToIndex
+                                                                 @flat-list-ref
```
Comment (clauxx) [inline src/quo/components/tabs/tabs/view.cljs]: `reagent/next-tick` could be used instead i think

File: src/quo/components/tabs/tabs/view.cljs (hunk 4 above)
Comment (clauxx) [reply src/quo/components/tabs/tabs/view.cljs]: why would it need to re-render?

File: src/quo/components/tabs/tabs/view.cljs (hunk 5 above)
Comment (clauxx) [reply src/quo/components/tabs/tabs/view.cljs]: yeah, it would be best to pass `tab-view` just a `on-press` prop instead of passing all of: `flat-list-ref`, `set-active-tab-id`, `scroll-on-press`, `on-change`.

File: src/quo/components/tabs/tabs/view.cljs (hunk 4 above)
Comment (flexsurfer) [reply src/quo/components/tabs/tabs/view.cljs]: because when it will be rendered first time @flat-list-ref will be nil

Comment (clauxx) [reply src/quo/components/tabs/tabs/view.cljs]: hmm yeah, usually `forwardRef` is used for this, but it's kinda unnecessary to pass the ref here anyways. Better just create a `on-press` handler in the parent and pass it down instead of the ref directly.

File: src/quo/components/tabs/tabs/view.cljs (hunk 6 above)
Comment (ilmotta) [reply src/quo/components/tabs/tabs/view.cljs]: I originally coded a Reagent atom as a map because I wasn't sure if the fading calculation would change in the future to be more customizable. There was some uncertainty from designers. I see now we don't need to have a map as the state.

File: src/quo/components/tabs/tabs/view.cljs [hunk 9]
```diff
@@ -72,31 +71,33 @@
...
-                            (when (and scroll-on-press? @flat-list-ref)
-                              (.scrollToIndex ^js @flat-list-ref
-                                              #js
-                                               {:animated     true
-                                                :index        index
-                                                :viewPosition 0.5}))
-                            (when on-change
-                              (on-change id)))}
-    label]])
+  (let [on-press (rn/use-callback (fn [tab-id]
```
Comment (ilmotta) [inline src/quo/components/tabs/tabs/view.cljs]: @ajayesivan I'm seeing we're using `use-callback` incorrectly. One of the many perils of dropping down to the React layer. And I believe we'll continue to see more and more of this problem because that's a common problem in React apps too. We're no different.

React's `useCallback` will cache the function for subsequent renders relying exclusively on the deps passed to the hook. When we don't pass all the necessary dependencies, the cached function, if used again, will be incorrect.

In this code, `index` and `scroll-on-press?` are not being passed as dependencies. This code's PR only works if `data` (the sequence of tabs) doesn't change its size ever, which is a fragile assumption to be made.

Here's a proof demonstrating a bug:

1. In the Activity Center, there are 9 tabs. 0 to 8 `index`.
2. Make the tabs in `status-im.contexts.shell.activity-center.tabs.view` dynamic with `use-state`.
3. Set a new tab state where you have only 2 tabs (e.g. `Mentions` and `Membership`)
4. The UI re-rendered and now shows only two tabs.
5. Press on any of the tabs and watch the app explode with `scrollToIndex out of range`.

One might say, "but wait, you're forcing a bug, the current prod code uses static tabs". For that I say no, React components should be made dynamic and render correctly whenever state changes. Here it's not even the component's state directly, but a prop change from a parent view, so it's worse. @ajayesivan, this is by no means a feedback for your PR alone, but a general problem we'll continue to see and we need to be careful.

@flexsurfer this is a good example of a problem we didn't have with Reagent (missing deps for cached functions).

-------

Long story short, `use-callback` requires careful thought from developers. It's not always needed, just as we coded for years with Reagent without caching all functions. Particularly in components that render only once or re-render rarely I would say it's actually detrimental to the quality of the app.

cc @ulisesmac this is one good example about what we were talking today about hooks' deps.

Comment (flexsurfer) [reply src/quo/components/tabs/tabs/view.cljs]: Devs must understand how react works, and we should have full control of it, just rerendering functions and components each time is not a good option, so reagent simply hid the problems by paying for it in performance

File: src/quo/components/tabs/tabs/view.cljs [hunk 10]
```diff
@@ -118,87 +119,99 @@
...
+        set-initial-scroll-poisition      (rn/use-callback (fn []
+                                                             (js/setTimeout
+                                                              (fn []
+                                                                (.scrollToIndex
+                                                                 @flat-list-ref
+                                                                 #js
+                                                                  {:animated false
+                                                                   :index
+                                                                   (utils.collection/first-index
+                                                                    #(= active-tab-id (:id %))
```
Comment (flexsurfer) [inline src/quo/components/tabs/tabs/view.cljs]: here probably data and active-tab-id should be in deps as well

File: src/quo/components/tabs/tabs/view.cljs (hunk 9 above)
Comment (ilmotta) [reply src/quo/components/tabs/tabs/view.cljs]: > Devs must understand how react works, and we should have full control of it, just rerendering functions and components each time is not a good option, so reagent simply hid the problems by paying for it in performance

Right @flexsurfer, we should all know what we're using 👍🏼 

My point is different, it's that `use-callback` shouldn't be used all the time because caching functions in React is only ever (maybe) going to make a difference in parts of the tree that re-render more frequently. And the downside of using it is that it's very easy to use the incorrect deps and cause hard to debug issues. As I've shown in my previous comment.

In React apps we have linters to help detect incorrect deps. UIx for instance does this check at _compile-time_ (currently experimental). We have little to no protection in status-mobile.

Overall, I think we shouldn't be using `use-callback` everywhere there's an anonymous function. Caching functions is not needed all the time, far from that, because the cache will (maybe) be used only on the next render (if it happens). Our guidelines should be flexible regarding `use-callback`.

In the guidelines:

```
rn/use-callback hook should be used for anon callback functions
```

In essence, this hard rule shouldn't always be followed in my experience with React apps.

Just cc'ing a few others to raise a bit more awareness: @J-Son89, @smohamedjavid, @mohsen-ghafouri

Comment (cammellos) [reply src/quo/components/tabs/tabs/view.cljs]: I agree here that the guideline is too strict. Whether we use reagent or no, the problem is the same, we want to prevent recreating function across re-renders in performance critical paths, or have the function cause unnecessary re-renders because of the way is defined. 

But using `rn/use-callback` for everything sounds like a sledgehammer and I think gives a false sense of security, like in the case above, as it might be unnecessary, and I agree that it can be prone to errors (long interesting discussion to be had, I think, it also forces developers to think about dependencies of the function, which I guess is good as many of the mistakes pointed out were made because we didn't think about dependencies, but not worth the candle imo, but that's a separate discussio) . 

Unfortunately it's one of those cases were developers needs to think about the lifecycle and dependencies of their components and write the code accordingly, if they are not doing that, that's were code review and more introspection on why something re-renders and how many times will help. 

So I would agree that the guideline need to be changed into something that will give developers more agency on when it's good to use, and when it's not necessary.

**Comment by pavloburykh** [comment]
Hey @ajayesivan! Could you please give an update regarding this PR, is it ready for testing or in progress? Thank you.

**Review by clauxx — APPROVED** [review]

**Review by flexsurfer — APPROVED** [review]

**Comment by ajayesivan** [comment]
@pavloburykh it's ready for testing.

**Comment by VolodLytvynenko** [comment]
Hi @ajayesivan thank you for PR. No issues from my side. PR is ready to be merged
